examples/mem-correctness-test/src/main.rs

```diff
@@ -0,0 +1,279 @@
+/* This test based on the following pseudo code and tests correctness of our
+ * new allocator snmalloc
+
+   for 0..num_threads {
+    loop {
+        let mut regions: Vec<(Box<[u8], u8>)>;
+        let mut mem_used = 0u64;
+        match rnd % 4 {
+           0 => // Check area
+           1..2 => // Alloc random area if less than x% of the heap is used (not 100% to
+                   // account for fragmentation, ...) and check area.
+           3 => // Free random area
+        }
+    }
+}
+ * So this basically runs for a long time and should never crash
+ */
+
+use core::arch::asm;
+use rand::Rng;
+use std::mem;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Condvar, Mutex};
+use std::thread;
+
+extern "C" {
+    static HEAP_BASE: u64;
+    static HEAP_SIZE: usize;
+}
```

Comment:
Pretty much all this code could also be executed on almost all targets the Rust compiler supports, except some statements/functions like these. You could restrict these to only the SGX target with:
```
#[cfg(target_env = "sgx")]
extern "C" {
    static HEAP_BASE: u64;
    static HEAP_SIZE: usize;
}
```
That way you can develop and test under Linux, and if that works, try it within an enclave.